plugins/hls-gadt-plugin/src/Ide/Plugin/GADT.hs

```diff
@@ -0,0 +1,135 @@
+    range <- ExceptT
+        $ pure
+        $ maybeToEither "Unable to get data decl range"
+        $ srcSpanToRange $ locA ann
+    pragma <- getNextPragma state nfp
+    let insertEdit = [insertNewPragma pragma GADTs | all (`notElem` exts) [GADTSyntax, GADTs]]
+
+    _ <- lift $ sendRequest
+            SWorkspaceApplyEdit
+            (ApplyWorkspaceEditParams Nothing (workSpaceEdit nfp (TextEdit range txt : insertEdit)))
+            (\_ -> pure ())
+
+    pure Null
+    where
+        workSpaceEdit nfp edits = WorkspaceEdit
+            (pure $ HashMap.fromList
+                [(filePathToUri $ fromNormalizedFilePath nfp,
+                 List edits)])
+                 Nothing Nothing
+
+codeActionHandler :: PluginMethodHandler IdeState TextDocumentCodeAction
+codeActionHandler state plId@(PluginId txt) (CodeActionParams _ _ doc range _) = response $ do
+    nfp <- getNormalizedFilePath plId doc
+    (inRangeH98Decls, _) <- getInRangeH98DeclsAndExts state range nfp
+    let actions = map (mkAction . printOutputable . tcdLName . unLoc) inRangeH98Decls
+    pure $ List actions
+    where
+        mkAction :: T.Text -> Command |? CodeAction
+        mkAction name = InR CodeAction{..}
+            where
+                _title = "Convert \"" <> name <> "\" to GADT syntax"
+                _kind = Just CodeActionRefactorRewrite
+                _diagnostics = Nothing
+                _isPreferred = Nothing
+                _disabled = Nothing
+                _edit = Nothing
+                _command = Just
+                    $ mkLspCommand plId toGADTSyntaxCommandId _title (Just [toJSON mkParam])
+                _xdata = Nothing
+
+        mkParam = ToGADTParams (doc ^. L.uri) range txt
+
+-- | Get all H98 decls in the given range, and enabled extensions
+getInRangeH98DeclsAndExts :: (Monad (t IO), MonadTrans t) =>
+    IdeState
+    -> Range
+    -> NormalizedFilePath
+    -> ExceptT String (t IO) ([LTyClDecl GP], [Extension])
+getInRangeH98DeclsAndExts state range nfp = do
+    pm <- handleMaybeM "Unable to get ParsedModuleWithComments"
+        $ lift
+        $ runAction "GADT.GetParsedModuleWithComments" state
+        $ use GetParsedModuleWithComments nfp
+    let (L _ hsDecls) = hsmodDecls <$> pm_parsed_source pm
+        decls = filter isH98DataDecl
+            $ mapMaybe getDataDecl
+            $ filter (inRange range) hsDecls
+        exts = (toList . extensionFlags . ms_hspp_opts . pm_mod_summary) pm
+    pure (decls, exts)
+
+-- Copy from hls-alternate-number-format-plugin
+getNextPragma :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m NextPragmaInfo
+getNextPragma state nfp = handleMaybeM "Error: Could not get NextPragmaInfo" $ do
+      ghcSession <- liftIO $ runAction "GADT.GhcSession" state $ useWithStale GhcSession nfp
+      (_, fileContents) <- liftIO $ runAction "GADT.GetFileContents" state $ getFileContents nfp
+      case ghcSession of
```

Comment:
We can't move this to hls-plugin-api unless hls-plugin-api has dependency of ghcide

Reply:
I think it would be good to move the pragma-insertion logic to `hls-plugin-api` from `ghcide`, since it's a very common thing to want to do. Ideally we could wrap it all up in something like `insertPragmaIfNotPresent` or something.

Reply:
Did you figure out whether there was anything sensible we can do to share code here? Or do you want to leave it as a follow-up?